src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+                    if (IsArgumentListAndParameterListPerfactMatch(semanticModel, argumentList.Arguments, methodSymbol.Parameters, targetArgument, cancellationToken, out var paramIndex))
+                    {
+                        var correspondingParameter = methodSymbol.Parameters[paramIndex];
+                        var argumentConversionType = correspondingParameter.Type;
+
+                        if (correspondingParameter.IsParams && correspondingParameter.Type is IArrayTypeSymbol arrayType && !(targetNodeType is IArrayTypeSymbol))
+                        {
+                            // target argument is matched to the parameter with keyword params
+                            argumentConversionType = arrayType.ElementType;
+                        }
+
+                        potentialConversionTypes.Add(argumentConversionType);
+                    }
+                }
+
+                // Sort the potential conversion types by inheritance distance
+                var comparer = new InheritanceDistanceComparer(semanticModel, targetNodeType);
+                potentialConversionTypes.Sort(comparer);
+                potentialConvTypes = potentialConversionTypes.Distinct().ToImmutableArray(); // clear up duplicate types
+                if (potentialConvTypes.Length != 1)
+                {
+                    return false;
+                }
+
+                targetNodeConversionType = potentialConvTypes[0];
+            }
+
+            if (targetNodeConversionType == null)
+            {
+                return false;
+            }
+
+            var commonConversion = semanticModel.Compilation.ClassifyCommonConversion(targetNodeType, targetNodeConversionType);
+            // example: Derived d = [||]new Base();
+            // It is invalid except the target node has explicit conversion operator 
+            if (targetNode.IsKind(SyntaxKind.ObjectCreationExpression) && !commonConversion.IsUserDefined)
+            {
+                targetNodeConversionType = null;
+                return false;
+            }
+            return commonConversion.Exists;
+        }
+
+        /// <summary>
+        /// Try to test if the invocation node is available to invoke the method.
+        /// </summary>
+        /// <param name="semanticModel"></param>
+        /// <param name="arguments">The arguments of invocation</param>
+        /// <param name="parameters">The parameters of </param>
+        /// <param name="targetArgument">The target argument that contains target node</param>
+        /// <param name="cancellationToken"></param>
+        /// <param name="targetParamIndex">Output the corresponding parameter index of the target arugment</param>
+        /// <returns>
+        /// True, if arguments and parameters match perfectly.
+        /// False, otherwise.
+        /// </returns>
+        private bool IsArgumentListAndParameterListPerfactMatch(SemanticModel semanticModel, SeparatedSyntaxList<ArgumentSyntax> arguments,
```

Comment:
This method looks scary to me. Historically, attempting to match arguments to parameters has been very fragile and hard to maintain code. We did add IOperation support to allow this matching for code without overload resolution errors, but you will not be able to use that because you have compile errors and overload resolution failure here. @CyrusNajmabadi do you know if we have any existing helpers in IDE layer to accomplish this?

Reply:
@RikkiGibson Hey Rikki, so sorry for disturbing you. Here is my function to do the matching between arguments and parameters. Looks scary, is it possible to have an IDE level API for it?

Reply:
Filed an issue for this function: https://github.com/dotnet/roslyn/issues/42149